### PR TITLE
GPII-2772: Windows system settings handler

### PR DESCRIPTION
**Pull request by stegru**
This is the settings handler that lets you perform changes using the same back-end as the system settings applet (aka, immersive control panel).

The main piece of work is performed in an external helper C# program, because it has to be running as 64-bit. This is built during `npm install` (via [provisioning/NpmInstall.ps1](https://github.com/stegru/windows/blob/GPII-2772/provisioning/NpmInstall.ps1#L17)).

### CODE REVIEW

**Comment by stegru** [comment]
```
default: WinRM transport: negotiate
Timed out while waiting for the machine to boot. This means that
Vagrant was unable to communicate with the guest machine within
the configured ("config.vm.boot_timeout" value) time period.
```

ok to test

**Comment by stegru** [comment]
another CI related error
```
Fetching upstream changes from https://github.com/stegru/windows.git
 > git --version # timeout=10
 > git fetch --tags --progress https://github.com/stegru/windows.git +refs/heads/*:refs/remotes/origin/*
ERROR: Timeout after 10 minutes
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --tags --progress https://github.com/stegru/windows.git +refs/heads/*:refs/remotes/origin/*" returned status code 143
```

ok to test

**Comment by stegru** [comment]
ok to test

**Comment by the-t-in-rtf** [comment]
So, the changes to the `package.json` to run the build steps as part of `npm install` are a key improvement, but now Vagrant builds are running that step twice.  Unless there is a key reason to do both, I'd suggest updating this line:

https://github.com/stegru/windows/blob/GPII-2772/Vagrantfile#L45

**Comment by stegru** [comment]
I've removed the bits of `Build.ps1` that are called in `npm install`.

`NpmInstall.ps1` gets calles from `npm install`, this generates the dependencies for gpii-windows (the test-window and helper executable that this PR is for).

**Comment by stegru** [comment]
ok to test

```
==> default: Destroying VM and associated drives...
/opt/vagrant/embedded/gems/gems/httpclient-2.8.3/lib/httpclient/session.rb:623:in `rescue in create_socket': Connection refused - Connection refused - connect(2) for "127.0.0.1" port 5985 (127.0.0.1:5985) (Errno::ECONNREFUSED)
```

File: settingsHelper/SettingsHelper/ISettingItem.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        int GetProperty(string name);
+        int SetProperty(string name, object value);
+
+        // For Type = Action - performs the action.
+        IntPtr Invoke(IntPtr a, Rect b);
+
+        // SettingChanged event
+        event EventHandler<string> SettingChanged;
+        //int add_SettingChanged(int v);
+        //int remove_SettingChanged();
```
[the remark this answers is not in the document]
Comment (stegru) [reply settingsHelper/SettingsHelper/ISettingItem.cs]: They're the methods that the above event is made from, not required.

File: settingsHelper/SettingsHelper/ISettingItem.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Runtime.InteropServices;
+
+    /// <summary>
+    /// Interface for the settings classes, instantiated by the GetSettings exports.
+    /// </summary>
+    /// <remarks>
+    /// Most of the information was taken from the debug symbols (PDB) for the relevent DLLs. The symbols
```
[the remark this answers is not in the document]
Comment (stegru) [reply settingsHelper/SettingsHelper/ISettingItem.cs]: Just added a few lines in the comment - it's not that complicated.

File: settingsHelper/SettingsHelper/Program.cs [hunk 3]
```diff
@@ -0,0 +1,167 @@
+﻿/*
```
[the remark this answers is not in the document]
Comment (stegru) [reply settingsHelper/SettingsHelper/Program.cs]: Yes

File: settingsHelper/SettingsHelper/SettingItem.cs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+            {
+                Thread.Sleep(delay);
+            }
+
+            return this.settingItem.GetValue(valueName);
+        }
+
+        private void SettingItem_SettingChanged(object sender, string s)
+        {
+            //Console.WriteLine(s);
```
[the remark this answers is not in the document]
Comment (stegru) [reply settingsHelper/SettingsHelper/SettingItem.cs]: Removed

File: settingsHelper/SettingsHelper/Program.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                first = false;
+
+                Result result = SettingHandler.Apply(payload);
+                Console.Write(result.ToString());
+            }
+
+            Console.Write("]");
+
+            // Wait for them all to complete.
+            const long Timeout = 5000;
```
[the remark this answers is not in the document]
Comment (stegru) [reply settingsHelper/SettingsHelper/Program.cs]: Settings are normally applied in less than a second. The timeout is only noticed if something failed. 5 seconds gives plenty of time to apply the setting, but isn't too long to wait if there's a problem.

File: settingsHelper/SettingsHelper/Program.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    using System.IO;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.Serialization;
+    using Microsoft.Win32;
+
+    /// <summary>
+    /// Defines the <see cref="Program" />
+    /// </summary>
+    public class Program
```
[the remark this answers is not in the document]
Comment (stegru) [reply settingsHelper/SettingsHelper/Program.cs]: Done

**Comment by stegru** [comment]
@amb26 - you're it.

File: settingsHelper/SettingsHelper/ISettingItem.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+        // Performs an action
+        Action = 7,
+
+        // Needs investigating
+        SettingCollection = 8,
+    }
+
+    /// <summary>
+    /// Used by ISettingsItem.Invoke (reason unknown).
```
Comment (amb26) [inline settingsHelper/SettingsHelper/ISettingItem.cs]: So hilarious

File: package.json [hunk 8]
```diff
@@ -50,5 +50,8 @@
     "main": "./gpii.js",
     "engines": {
         "node": ">=6.3.1"
+    },
+    "scripts": {
+        "install": "powershell  –ExecutionPolicy Bypass provisioning/NpmInstall.ps1"
```
Comment (amb26) [inline package.json]: This character here doesn't actually appear to be a hyphen - what is it?

File: gpii/node_modules/systemSettingsHandler/src/systemSettingsHandler.js [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+var windows = fluid.registerNamespace("gpii.windows");
+
+fluid.registerNamespace("gpii.windows.systemSettingsHandler");
+
+/**
+ * Executes the settings helper application.
+ *
+ * See the documentation for the application for information about what is send and received.
+ *
+ * @param settings {object} The JSON to pass to the application. An array of objects containing settingID, method, and
```
Comment (amb26) [inline gpii/node_modules/systemSettingsHandler/src/systemSettingsHandler.js]: Wrong way round for our new Linting rules (coming soon!) should read @param {Object} settings - The JSON

Please add further elements
   {String} settings.method - The method etc.

File: gpii/node_modules/systemSettingsHandler/src/systemSettingsHandler.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+ *      "oldValue": {
+ *        "value": true,
+ *      },
+ *      "newValue": {
+ *        "value": true
+ *      }
+ *    }
+ *    "SystemSettings_Accessibility_Magnifier_IsEnabled": { ... }
+ *  }
+ * @param payload {object} The payload.
```
Comment (amb26) [inline gpii/node_modules/systemSettingsHandler/src/systemSettingsHandler.js]: Fix globally - @param {Object} payload 
@returns -> @return

File: gpii/node_modules/processHandling/test/test-window.cs [hunk 11]
```diff
@@ -0,0 +1,73 @@
+/*
+ * test-window.c: Simple application that creates a window, used to test closeProcessByName.
+ *
+ * How to build:
+ *  csc test-window.cs
+ *
+ * Copyright 2016 Raising the Floor - International
```
Comment (amb26) [inline gpii/node_modules/processHandling/test/test-window.cs]: That was, in fact, the previous century, Archchancellor

File: package.json [hunk 12]
```diff
@@ -50,5 +50,8 @@
     "main": "./gpii.js",
```
Comment (amb26) [inline package.json]: I've issued a new universal dev build from current trunk 0.3.0-dev.20180427T101505Z.dad9bfb - could you update to that and check it still works, cheers

**Comment by amb26** [comment]
Ok oops, well at least we have learned clearly that it does not work. I've pinged @kaspermarkus to take a look at his Windows pull - in the meantime let's drop back to the previous release

File: package.json (hunk 8 above)
Comment (amb26) [reply package.json]: When you're viewing diffs on a Windows 7 command prompt with no Unicode fonts, these things are rather easy to spot :)

File: package.json (hunk 12 above)
Comment (stegru) [reply package.json]: It looks like that universal requires https://github.com/GPII/windows/pull/120

**Comment by stegru** [comment]
I've updated the universal reference to use https://github.com/GPII/universal/pull/594. It contains a preference set `nyx` which uses this settings handler.

**Comment by stegru** [comment]
```
Failures
 - python2 (exited 1) - python2 not installed. An error occurred during installation:
 The operation has timed out
VERBOSE: 'C:\ProgramData\chocolatey\bin\choco.exe' exited with 1
```

ok to test

**Comment by stegru** [comment]
ok to test (big update to universal)

File: package.json [hunk 13]
```diff
@@ -11,13 +11,14 @@
...
-        "posttest": "node_modules/.bin/nyc report -r html -r text-summary"
+        "posttest": "node_modules/.bin/nyc report -r html -r text-summary",
+        "install": "powershell -ExecutionPolicy Bypass provisioning/NpmInstall.ps1"
     },
     "dependencies": {
         "edge": "6.5.1",
         "edge-js": "8.8.1",
         "ffi": "2.0.0",
-        "gpii-universal": "0.3.0-dev.20180608T130946Z.7218a34",
+        "gpii-universal": "stegru/universal#GPII-2717",
```
Comment (amb26) [inline package.json]: I assume you meant this to be bound to your branch #GPII-2772 which is the matching one from https://github.com/GPII/universal/pull/594 - GPII-2717 looks to be a very stale branch

Comment (stegru) [reply package.json]: no wonder the test passed!

**Comment by amb26** [comment]
Pretty weird integrity failure:
````
default: npm ERR! code EINTEGRITY
    default: npm ERR! sha1-0Ng2s0OqbbiMDez+rnHmJ11BQg4= integrity checksum failed when using sha1: wanted sha1-0Ng2s0OqbbiMDez+rnHmJ11BQg4= but got sha1-kZpOsrsMqUlqMV9U2uFSFQ36qa0=. (7551734 bytes)
````
npm's shoddy error reporting doesn't make it clear which package. But a big one ....

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
Proper failure this time, pretty worrying. It may be random, given I did verify that this worked locally: 
````
09:03:26.245:  jq: Expected: {
    "gpii.windows.enableRegisteredAT": {
        "com.microsoft.windows.magnifier": [
            {
                "settings": {
                    "running": true
                }
            }
        ],
        "com.microsoft.windows.onscreenKeyboard": [
            {
                "settings": {
                    "running": true
                }
            }
        ]
    }
}
09:03:26.249:  jq: Actual: {
    "gpii.windows.enableRegisteredAT": {
        "com.microsoft.windows.magnifier": [
            {
                "settings": {
                    "running": true
                }
            }
        ],
        "com.microsoft.windows.onscreenKeyboard": [
            {
                "settings": {
                    "running": false
                }
            }
        ]
    }
}
````

**Comment by amb26** [comment]
ok to test

**Comment by amb26** [comment]
Terrible!

**Pull request merged**